src/logic/embedding.lean

```diff
@@ -320,3 +320,57 @@ namespace set
 ⟨λ x, ⟨x.1, h x.2⟩, λ ⟨x, hx⟩ ⟨y, hy⟩ h, by { congr, injection h }⟩
 
 end set
+
+section subtype
+
+variable {α : Type*}
+
+def subtype_or_embedding (p q : α → Prop) [decidable_pred p] :
+  {x // p x ∨ q x} ↪ {x // p x} ⊕ {x // q x} :=
+⟨λ x, if h : p x then sum.inl ⟨x, h⟩ else sum.inr ⟨x, x.prop.resolve_left h⟩,
+  begin
+    intros x y,
+    dsimp only,
+    split_ifs;
+    simp [subtype.ext_iff]
+  end⟩
+
+lemma subtype_or_embedding_apply_left {p q : α → Prop} [decidable_pred p] (x : {x // p x ∨ q x})
+  (hx : p x) : subtype_or_embedding p q x = sum.inl ⟨x, hx⟩ := dif_pos hx
+
+lemma subtype_or_embedding_apply_right {p q : α → Prop} [decidable_pred p] (x : {x // p x ∨ q x})
+  (hx : ¬ p x) : subtype_or_embedding p q x = sum.inr ⟨x, x.prop.resolve_left hx⟩ := dif_neg hx
+
+@[simps] def subtype.imp_embedding (p q : α → Prop) (h : p ≤ q) :
+  {x // p x} ↪ {x // q x} :=
+⟨λ x, ⟨x, h x x.prop⟩, λ x y, by simp [subtype.ext_iff]⟩
+
+def subtype_or_equiv (p q : α → Prop) [decidable_pred p] (h : disjoint p q) :
+  {x // p x ∨ q x} ≃ {x // p x} ⊕ {x // q x} :=
```

Comment:
Isn't this easier to build out of `equiv.sum_compl` and `equiv.subtype_univ_equiv`?

Reply:
I am not sure how to do this precisely, since all of the subtypes here are over the base type, and there is no requirement that `p x ∨ q x` is true for all `x : α`.

Reply:
Oh, you're right.